### ISSUE TITLE
[Zamba2] Causality violation in torch_forward: inter-chunk state recurrence reduces over the wrong axis

### ISSUE THREAD

**Issue by openaiteams** [huggingface/transformers#47475]
### System Info

- transformers version: 5.6.0.dev0
- Platform: Linux (x86_64)
- Python version: 3.10
- PyTorch version: 2.10.0+cu126
- Device used for this report: CPU
- mamba_ssm / causal_conv1d: not installed, so the pure-PyTorch path (torch_forward) is taken

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Reproduction

In the pure-PyTorch path of Zamba2, a token at position P changes the logits at positions before P, which violates autoregressive causality.
```python
import torch
from transformers import AutoModelForCausalLM
m = AutoModelForCausalLM.from_pretrained(
    "Zyphra/Zamba2-1.2B", dtype=torch.float32
).eval()
torch.manual_seed(0)                 # seed after loading (loading consumes RNG)
T, P = 512, 400                      # sequence length 512, perturb position 400
ids  = torch.randint(5, 1000, (1, T))
ids2 = ids.clone()
ids2[0, P] = (ids2[0, P] + 7) % 1000 + 5     # change ONLY the token at position 400
with torch.no_grad():
    a = m(ids,  use_cache=False).logits
    b = m(ids2, use_cache=False).logits
# positions 0..399 precede the change, so they must be identical
print((a[:, :P] - b[:, :P]).abs().max().item())
```
Output:
```
1.1241e-02        # expected: 0.0
```
Looking at where the divergence starts:
```
pos   0 : 0.0000e+00
pos 100 : 0.0000e+00
pos 200 : 0.0000e+00
pos 255 : 0.0000e+00     <- clean up to here
pos 256 : 5.7473e-03     <- contamination starts exactly here
pos 300 : 5.0647e-03
pos 399 : 3.2187e-03
```
Contamination begins at position 256, which is exactly `config.chunk_size`.
**Conditions for the bug to appear**
| Condition | Detail |
|---|---|
| Code path | The pure-PyTorch path, taken when `mamba_ssm` / `causal_conv1d` are not installed |
| Sequence length | Must exceed `chunk_size` (256). At `T <= 257` there is only one chunk and nothing is observed |
| Weights | Irrelevant - reproduces identically with a randomly initialized model built via `from_config` |
The last row matters: this is not specific to any checkpoint.
Note that the least significant digits of the magnitude vary with the execution environment (e.g. CPU thread count). What is invariant is that contamination starts exactly at `chunk_size` and every position before it is exactly `0`.
## Root cause
`segment_sum` applies a lower-triangular mask, so `decay_chunk[b, h, i, j]` is only valid for `j <= i`, where `i` indexes the output chunk and `j` the input chunk. The recurrence must therefore reduce over the input-chunk axis `j`.
`models/mamba2/modeling_mamba2.py` (correct):
```python
decay_chunk = decay_chunk.transpose(1, 3)                                          # (b,h,i,j) -> (b,j,i,h)
new_states = (decay_chunk[..., None, None] * states[:, :, None, ...]).sum(dim=1)   # reduce over j
```
`models/zamba2/modeling_zamba2.py` (buggy):
```python
states_permuted = states.permute(0, 2, 1, 3, 4)
result = (decay_chunk[..., None, None] * states_permuted[:, :, None, ...]).sum(dim=2)   # reduces over i
new_states = result.permute(0, 2, 1, 3, 4)
```
Without the `transpose(1, 3)`, reducing over `dim=2` sums over the output-chunk axis. Because of the lower-triangular mask only terms with `i >= j` survive, so chunk `j` receives contributions from later chunks.
`decay` derives from `A_cumsum`, which derives from `dt`, which depends on the input. Future tokens therefore influence earlier outputs, consistent with the contamination starting exactly at a chunk boundary.
## Proposed fix
In `Zamba2MambaMixer.torch_forward`, replace the three lines with the Mamba2 formulation:
```diff
-            states_permuted = states.permute(0, 2, 1, 3, 4)
-            result = (decay_chunk[..., None, None] * states_permuted[:, :, None, ...]).sum(dim=2)
-            new_states = result.permute(0, 2, 1, 3, 4)
+            decay_chunk = decay_chunk.transpose(1, 3)
+            new_states = (decay_chunk[..., None, None] * states[:, :, None, ...]).sum(dim=1)
```
## Verification
Applying only the patch above, with the same input:
| | max abs delta over positions 0..399 | Contamination onset |
|---|---|---|
| Original | ~1.12e-02 | 256 |
| Patched | 0.0000e+00 | none |
The patch fully restores causality, and no other source of leakage remains.
I am happy to open a PR with this change if that is useful.

### Expected behavior

`(a[:, :P] - b[:, :P]).abs().max()` should be `0.0`: changing the token at position `P` must not affect any output at positions `< P`.
